File: 리트코드/0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        int size = nums.length;
        if (size == 0) {
            return null;
        }
        return construct(nums, 0, size - 1);
        
    }

    public TreeNode construct(int[] nums, int lo, int hi) {
        if (hi < lo) {
            return null;
        }
        int mid = lo + (hi - lo) / 2;

        TreeNode node = new TreeNode(nums[mid]);

        node.left = construct(nums, lo, mid -1);
        node.right = construct(nums, mid + 1, hi);

        return node;
    }
}

File: 리트코드/0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun sortedArrayToBST(nums: IntArray): TreeNode? {
        fun construct(lo:Int, hi:Int): TreeNode?{
            if(hi < lo) return null
            val mid = lo + (hi-lo) / 2

            val node:TreeNode = TreeNode(nums[mid])
            node.left = construct(lo,mid - 1)
            node.right = construct(mid + 1, hi)
            
            return node
        }

        return construct(0,nums.size-1)
    }
    
    
}